Move and rewrite intro pattern section (#9269)
coq/coq: doc/sphinx/proof-engine/tactics.rst

tactics. This chapter is devoted to atomic tactics. The tactic
language will be described in Chapter :ref:`ltac`.

Common elements of tactics
--------------------------

.. _invocation-of-tactics:

Invocation of tactics

> ~~~~~~~~~~~~~~~~~~~~~

Reviewer: You have missed a leading `|` just below in the "tactic invocation" grammar.

tactics. This chapter is devoted to atomic tactics. The tactic
language will be described in Chapter :ref:`ltac`.

Common elements of tactics
--------------------------

.. _invocation-of-tactics:

Invocation of tactics
~~~~~~~~~~~~~~~~~~~~~

A tactic is applied as an ordinary command. It may be preceded by a
goal selector (see Section :ref:`ltac-semantics`). If no selector is
specified, the default selector is used.

.. _tactic_invocation_grammar:

  .. productionlist:: sentence
     tactic_invocation : `toplevel_selector` : `tactic`.

>                        : `tactic` .

Reviewer: ```suggestion
                       : `tactic`.
```

Bindings list
~~~~~~~~~~~~~~~~~~~

Tactics that take a term as an argument may also support a bindings list,
so as to instantiate some parameters of the term by name or position.
The general form of a term equipped with a bindings list is ``term with

> bindings_list`` where ``bindings_list`` may be of two different forms:

Reviewer: Should use ``` :n:`@term with @bindings_list` where :token:`bindings_lists`... ```

.. _bindings_list_grammar:

  .. productionlist:: `bindings_list`
     bindings_list : (`ref` := `term`) ... (`ref` := `term`)
                   : `term` ... `term`

> + In a bindings list of the form :n:`{+ (@ref:= @term)}`, :n:`@ref` is either an

Reviewer: If `ref` was never defined in a grammar, it would be worth adding it to the grammar above.

Author: I added:

```
     ref           : ident
                   : INT
```

And I noticed that "INT" is a poor name, since it only includes natural numbers.  Maybe I'll clean that up in another PR.

  :n:`@ident` or a :n:`@num`. The references are determined according to the type of

>   ``term``. If :n:`@ref` is an identifier, this identifier has to be bound in the

Reviewer: ```suggestion
   :n:`@term`. If :n:`@ref` is an identifier, this identifier has to be bound in the
```

Same change to do three other times in this paragraph, and one missing `@` in the next paragraph.

Bindings list
~~~~~~~~~~~~~~~~~~~

Tactics that take a term as an argument may also support a bindings list
to instantiate some parameters of the term by name or position.
The general form of a term with a bindings list is
:n:`@term with @bindings_list` where :token:`bindings_list` can take two different forms:

.. _bindings_list_grammar:

  .. productionlist:: bindings_list
     ref           : ident

>                    : INT

Reviewer: Change this in:

```rst
  .. productionlist:: bindings_list
     ref           : `ident`
                   : `num`
```

`num` and `ident` are both already defined grammar tokens.

     bindings_list : (ref := `term`) ... (ref := `term`)
                   : `term` ... `term`

> + In a bindings list of the form :n:`{+ (ref:= term)}`, :n:`ref` is either an

Author: I think it's OK.  "bindings list" can only mean a list of bindings.  "binding list" could mean "list of bindings" or "binding" could be taken as a modifier of "list" (like "binding agreement" or "sorted list").  The phrase appears in 9 places in the file, BTW.

  .. exn:: Not the right number of missing arguments.
     :undocumented:

.. _intropatterns:

Intro patterns
~~~~~~~~~~~~~~

Intro patterns let you specify the name to assign to variables and hypotheses
introduced by tactics.  They also let you split an introduced hypothesis into
multiple hypotheses or subgoals.  Common tactics that accept intro patterns

> include :tacn:`intros`, :tacn:`destruct` and :tacn:`apply`.

Reviewer: Yes, but I think we should have more cross-links between the Tactics chapter and the SSR chapter.

> include :tacn:`intros`, :tacn:`destruct` and :tacn:`apply`.

Author: I think such links should be considered as a group and done consistently after the tactics and SSR chapters have been extensively revised.  I don't want to consider that now.

> .. productionlist::        coq

Author: "The argument to productionlist serves to distinguish different sets of production lists that belong to different grammars."

> .. productionlist::        coq

Reviewer: Interesting, it's not how it was used so far in the manual...

> .. productionlist::        coq

Author: I expect that needs untangling at some point.  `productionlist:: coq` is used 9 time in the .rst files.

   simple_intropattern     : `simple_intropattern_closed`

>                            : % `operconstr` ... % `operconstr`

Author: It's defined as: ```LIST0 ["%"; c = operconstr LEVEL "0" -> { c } ]"``` in simple_intropattern.

>                            : % `operconstr` ... % `operconstr`

Reviewer: BTW you should probably change this `operconstr` into `term` as it seems we rarely distinguish between the various term representations in the grammars in the manual.

>                            : % `operconstr` ... % `operconstr`

Author: I'd like to understand that better first.  "term" is not even a non-terminal in the grammar.

>                            : % `operconstr` ... % `operconstr`

Reviewer: The internal name for "term" is "constr" in Coq for historical reason (when Coq was called "The Construction Engine" from the "Calculus of Constructions"). Developers have wondered whether it should be renamed to "term" but it is used too much for such a change to be worthwhile... However, in the manual we use "term" instead to be less confusing.

>                            : % `operconstr` ... % `operconstr`

Reviewer 2: The PR is still evolving, so maybe this is not yet the right time to comment. A few remarks though.

Yes, `operconstr` is the internal parser name of the recursive grammar of terms. This grammar has two useful entry points: `constr` which is `operconstr` at level 8 (to be used in positions where applications needs to be parenthesized to avoid collisions with the rest of the grammar), and `lconstr` which is `operconstr` at level 200 (to be used in a context where the expression is already surrounded by delimiters).

We could in principle be explicit in the reference manual on the different uses of level, but it has been considered simple to call `term` any of `constr`, `lconstr` and `operconstr`. Some improvements could be done though.

   simple_intropattern_closed : `or_and_intropattern`
                              : `equality_intropattern`
                              : _
                              : `naming_intropattern`

>    or_and_intropattern     : [ `intropatterns` | ... | `intropatterns` ]

Author: I discovered that you don't have to have '|' to begin the 2nd..nth productions for a nonterminal.  I think we should remove them throughout the documentation because they are visual clutter, they make it harder to read and can confuse.

![image](https://user-images.githubusercontent.com/1253341/50367194-63f54480-0533-11e9-8028-679d3cc2a688.png)

Reviewer: Sounds good.

                           : ()
                           : ( `simple_intropattern` , ... , `simple_intropattern` )
                           : ( `simple_intropattern` & ... & `simple_intropattern` )
   equality_intropattern   : ->
                           : <-
                           : [= `intropatterns` ]
   intropatterns           : `nonsimple_intropattern` ...

>                            : <empty>

Author: It represents the empty production.  Otherwise it would just be an empty line, which could be interpreted as just a blank line.

![image](https://user-images.githubusercontent.com/1253341/50367732-0d8a0500-0537-11e9-90b7-3683c664df9c.png)

>    ne_intropatterns        : `nonsimple_intropattern` ...

Reviewer: "ne" like non-empty? This whole business of the "empty" intro-pattern seems puzzling to me.

Author: Probably related to this odd-looking code.  Gotta have the `ne_intropatterns` for the parser.  Also seems like `eintros` should be usable without intropatterns.

```
       IDENT "intros"; pl = ne_intropatterns ->
          { TacAtom (CAst.make ~loc @@ TacIntroPattern (false,pl)) }
      | IDENT "intros" ->
          { TacAtom (CAst.make ~loc @@ TacIntroPattern (false,[CAst.make ~loc @@IntroForthcoming false])) }
      | IDENT "eintros"; pl = ne_intropatterns ->
          { TacAtom (CAst.make ~loc @@ TacIntroPattern (true,pl)) }
```

Reviewer: I don't know if it's possible to simplify the code of the parser here, but at least the grammar in the manual should probably not mention this all empty / non-empty business.

   nonsimple_intropattern  : `simple_intropattern`
                           : *
                           : **

>    naming_intropattern     : `pattern_ident`

Reviewer: I would also prefer if the `pattern_ident` indirection were removed. It is only making the grammar longer with no benefit (of course, once again this is better if it can be removed in the code as well).

                           : ?
                           : `ident`
   pattern_ident           : ? `ident`

> 

Author: I thought about whether a) to show all these productions together, b) to show a few productions at a time just before the text describing their semantics and c) both a and b.  The grammar in its current form here would be confusing for option b or c.

Use these elementary patterns to specify a name:

* :n:`@ident` - use the specified name
* :n:`?` - let Coq choose a name

> * :n:`? @ident` - generate a name that begins with :n:`@ident`

Reviewer: ```suggestion
* :n:`?@ident` - generate a name that begins with :n:`@ident`
```

There should be no space, otherwise it will be parsed as two intropatterns. I suppose the grammar should be fixed as well to reflect that.

* :n:`_` - discards the matched part (unless it is required for another
  hypothesis)
* if a disjunction pattern omits a name, such as :g:`[|H2]`, Coq will choose a name

> 

Author: I wanted to use boldface for, say ```:n:`?` ``` but ```**:n:`?`**``` doesn't interpret the `:n:`.

The most common splitting patterns are:

* split a hypothesis in the form :g:`A /\ B` into two

>   hypotheses :g:`H1: A` and :g:`H2: B` using the pattern :g:`[H1 H2]`

Reviewer: I usually recommend using `(H1, H2)` instead in this case...

Author: Why not always `(H1 & H2)`?

Reviewer: Why not indeed? Although, if you have a left-associative operator (quite infrequent I suppose), you might need to use the `((H1, H2), H3)` which might still be preferred to `[[H1 H2] H3]`...

Reviewer 2: In the category of left-associative conjunctive operators, we missed the opportunity to make `prod` right-associative... (Too late, unless using a translator.)

Reviewer: This can be done in stdlib2 though.

* split a hypothesis in the form :g:`A \/ B` into two
  subgoals using the pattern :g:`[H1|H2]`.  The first subgoal will have the hypothesis
  :g:`H1: A` and the second subgoal will have the hypothesis :g:`H2: B`.

> **:ref:`Example <_intropattern_conj_ex>` doesn't work for linking to the example**

Reviewer: Did you try removing the leading underscore?

Author: That works, thanks.  (But too much arcana in rst, I think.)

**Sphinx says: "WARNING: undefined label: _intropattern_conj_ex (if the link has no caption
the label must precede a section header)"**

> 

Author: I wanted to put the word "Example" in each of the two bullets so the user could find the example immediately rather than scrolling down several pages.  I suppose I could make each example a sub-sub-sub section to get around the problem.  People might even like to see that in the table of contents (though it will make it bigger).

Splitting patterns can have more than 2 names, for example :g:`[H1 H2 H3]`

> matches :g:`A /\ B /\ C`.  Patterns can be nested:

Reviewer: That's incorrect. Because of the implicit parenthesizing the actual pattern used to destruct this is `[H1 [H2 H3]]`, alternatively `(H1, (H2, H3))` or the much nicer (and thus recommended) `(H1 & H2 & H3)` that is made just for that purpose.

Author: The & form is recommended over the , form?  Both the [ ] and the , form are not recommended (maybe deprecated someday?)

Reviewer: The point of the form with the comma is to look like a pair when destructuring something that can be understood as a pair. However; the intro-pattern with square brackets is more generic. The following cannot be done with any other intro pattern:

```
Inductive foo := | bla : nat -> nat -> foo | bar : nat -> foo.
Goal foo -> True.
intros [n1 n2|n3].
```

:g:`[[Ha|Hb] H]` can be used to split :g:`(A \/ B) /\ C`.

Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,

> use the :tacn:`split` tactic to replace the current goal with subgoals :g:`A` and :g:`B`.

Reviewer: Indeed, even if we had the idea: see #410.

> For a goal :g:`A \/ B`, use :g:`apply or_introl` to replace the current goal with :g:`A`, or

Reviewer: ```suggestion
For a goal :g:`A \/ B`, use :tacn:`left` to replace the current goal with :g:`A`, or
```

> :g:`apply or_intror` to replace the current goal with :g:`B`.

Reviewer: ```suggestion
:tacn:`right` to replace the current goal with :g:`B`.
```

> The other splitting patterns described below are not frequently used.

Reviewer: On the contrary, as I said before, the once with commas and `&` are very much recommended.

Reviewer: Indeed, but the code of the standard library is not a good example of how people would write things today because it is mostly ancient code that has very rarely been updated.

These patterns can be used when the hypothesis is an equality:

> * :n:`->` - replaces the right-hand side of the hypothesis with the left-hand

Reviewer: I just noticed that you sometimes put "s" at the end of this verbs and sometimes not.

Author: That's standard usage, right?  "[It] replaces ..." for a singular subject and "to replace" for the infinitive.  I didn't see a misuse.

Reviewer: For instance:

```
* :n:`? @ident` - generate a name that begins with :n:`@ident`
* :n:`_` - discards the matched part (unless it is required for another
  hypothesis)
```

  side of the hypothesis in the conclusion of the goal; the hypothesis is
  cleared; if the left-hand side of the hypothesis is a variable, it is
  substituted everywhere in the context and the variable is removed.
* :n:`<-` - similar to :n:`->`, but replaces the left-hand side of the hypothesis
  with the right-hand side of the hypothesis.
* :n:`[= @intropatterns ]` - If the product is over an equality type,
  applies either :tacn:`injection` or :tacn:`discriminate`
  instead of :tacn:`destruct`; if :tacn:`injection` is applicable, the patterns
  :n:`{+, p}` are used on the hypotheses generated by :tacn:`injection`; if the
  number of patterns is smaller than the number of hypotheses generated, the
  pattern :n:`?` is used to complete the list.

> 

Author: I didn't understand this, just copied the text with minor tweaks.

**need example for [= ... ]**

* :n:`( @simple_intropattern , ... , @simple_intropattern )` - matches
  a product over an inductive type :g:`I` with a single
  constructor that itself has at least `n` arguments.  It performs a case
  analysis over the hypothesis, as :n:`destruct` would, and applies the
  patterns to the arguments of the constructor of :g:`I`.  Note that

>   that :n:`({+ p})` is equivalent to :n:`[{+ p}]`);

Reviewer: ```suggestion
  that :n:`({+, @simple_intropattern})` is equivalent to :n:`[{+ @simple_intropattern}]`);
```

Reviewer: I suggest we stick to BNF only in the grammars themselves and inside, `:n:` blocks, we use these "diagrams" but with actual tokens when we can.

>   that :n:`({+ p})` is equivalent to :n:`[{+ p}]`);

Author: This is exactly equivalent to [A B C ...]?  If so, I would make it one sentence. The description is impenetrable.

Reviewer 2: The description is not up-to-date. It should be ``... constructor that itself has `n` arguments`` and it is not any longer equivalent. A correct formulation, hoping it is not too terse, could be ``Note that :n:`({+, p})` is equivalent to :n:`[{+ p}]` for a list made of as many patterns as the constructor has arguments`.

Note that in this case, there is a tolerance. The count of arguments is done both including or excluding let-ins and both cases are allowed. I.e., in:
```
Inductive I := C : forall a:nat, let b:=a in forall c:nat, I.
Goal forall x:I, x = x.
```
both `intros (a,b,c)` and `intros (a,c)` are allowed, and the latter is the same as `intros (a,?,c)`.
```

* :n:`( @simple_intropattern & ... & @simple_intropattern )` - :n:`({+& p})`
  is a shortcut for matching with
  :n:`(p,( ... ,( ..., p ) ... ))`.  It expects the hypothesis to be a sequence of
  right-associative binary inductive constructors such as :g:`conj` or
  :g:`ex_intro`; for instance, a hypothesis with type
  :g:`A /\(exists x, B /\ C /\ D)` can be introduced via pattern

>   :n:`(a & x & b & c & d)`;

Author: Couldn't get the example to work.  A generalization of [A B C]?  Didn't fully understand.

**need example for & ... &**

> **A /\\ (exists x, B /\\ C /\\ D) is not valid for (a & x & b & c & d)**

Reviewer: How did you tested it? This works:

```coq
Parameters A B C D : Prop.
Goal A /\ (exists x : nat, B /\ C /\ D) -> True.
intros (a & x & b & c & d).
```

Reviewer: There's no difference but you forgot to introduce the parameters in your context first, i.e. it should have been:

```
intros A x B C D (a & x & b & c & d).
```

Author: OK.  I didn't expect to see the `_` in this example.  The example would be simpler without that detail, but maybe that's the way it has to be.

```
Goal forall (A: Prop) (B: Prop) (C: Prop), A /\ (exists x:nat, B /\ C) -> True.
intros *.

1 subgoal
A, B, C : Prop
______________________________________(1/1)
A /\ (exists _ : nat, B /\ C) -> True

```

gives

Reviewer: Here it is, tweaked so that it does not display any `_`:

```
Goal forall (A: Prop) (B: nat -> Prop) (C: Prop), A /\ (exists x:nat, B x /\ C) -> True.
intros *.
intros (a & x & b & c).
```

> * :n:`*` - introduces one or more quantified variables from the result

Reviewer: ... until there are no more quantified variables

* :n:`**` - introduces one or more quantified variables or hypotheses from the result until there are
  no more quantified variables or implications (:g:`->`).  :g:`intros **` is equivalent
  to :g:`intros`.

* :n:`%term` - a pattern :n:`p{+ %term}` first applies :n:`{+ term}`
  on the hypothesis to be introduced (as in :n:`apply {+, term}`) prior to the

>   application of the introduction pattern :n:`p`;

Reviewer: Note in passing. This would be nice to use actual tokens more systematically instead of `p`

  the on-the-fly application of lemmas: :n:`p{+ %term}` where :n:`p`
  itself is not a pattern for on-the-fly application of lemmas (note:
  syntax is in experimental stage)  **huh?**

> **need example, maybe simplify "intros [= -> %IHl]." from List.v?**

Reviewer 3: Could this [example](https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v) be of any help here?

Author: Thanks, I'll adapt that.  But without some explanation, understanding it takes a couple minutes of experimentation:
- `intros` would just add `H : S (length ys) = 1`
- `intros [=]` would reduce (right term?) to add `H0 : length ys = 0`
- `intros [=->%length_zero_iff_nil]` applies the resulting equality `l=[]`

I wish Coq had a way to show these multiple substeps for this example (and many, many others).  It's not so easy to understand what Coq is doing when there are multiple substeps.  I think this would be very helpful for learners.  But creating such a feature might be difficult but maybe feasible.  Certainly I'd have to learn far more about tactic internals.  (What do you think?)

It may be helpful to have a way to do diffs that comparing the result of applying 2 alternative tactics.  (Maybe).

Off topic: have you tried the diffs feature in master?  It will be in 8.9, but it looks like 2 needed bug fix PRs for it may not make it into 8.9 :-(.

Reviewer: The `Info` command was an attempt to display such a decomposition but it is unfinished.

Reviewer 3: > I wish Coq had a way to show these multiple substeps for this example (and many, many others).

I fully agree. Although, I'd say that stepping through a proof (with all those `;`) without dismantling it first would be a the most desirable feature.

> Off topic: have you tried the diffs feature in master?

Not yet. Does ProofGeneral support it?

Author: @anton-trunov No, diffs are not in PG yet.  I was hoping someone who's worked on PG would volunteer to do the small amount of work needed for diffs in PG once 8.9 is out.  Perhaps unrealistic, I'll see.

> I'd say that stepping through a proof (with all those ;) without dismantling it first would be a the most desirable feature.

Maybe it would not be hard to add an option that would treat ; as if it was "." for non-LTAC code.   I suspect it may be complex to robustly extend the concept to LTAC, which I don't know well.   Would an option that doesn't work in LTAC still be worthwhile?

Reviewer: You cannot simply replace `;` with `.` and hope for the code to work though unless you add the proper goal selectors.

**operconstr: coqtop is not accepting this, not sure why--or what the construct means**
**is it experimental?**

.. note::

   In :n:`intros @intro_pattern_list`, if the last introduction pattern
   is a disjunctive or conjunctive pattern
   :n:`[{+| @intro_pattern_list}]`, the completion of :n:`@intro_pattern_list`
   so that all the arguments of the i-th constructors of the corresponding

>    inductive type are introduced can be controlled with the following option:

Author: Couldn't tell if this is important or not, or if it could be said more simply.

Reviewer: I just noted that this option, as its name suggests only impacts the last intro pattern. In the other cases, it has no effect and does not make `[]` behave like in SSReflect. I'm wondering what was the objective in introducing this option and whether we should remove it (simpler language, less options, less confusion).

Reviewer 2: The flag `Bracketing Last Introduction Pattern` is indeed a compatibility flag. It would probably be less confusion to eventually deprecate it.

Note that ssreflect introduction patterns extended an original mechanism of introduction patterns (in Coq 6.2) which was "procedural". For better compositionality, the introduction patterns evolved in vanilla Coq into a "structured" notion of introduction patterns but ssreflect kept the initial implementation of Coq 6.2. When the pattern is complete, in the sense of giving an introduction pattern for each of the argument of the constructor, there is no difference. There is a difference when it is incomplete. The original 6.2 implementation is such that, e.g., `[ | ] H` is dependent on the inductive structure of the first hypothesis. For instance, if the goal is `forall n:nat, n = 0 -> 0 = n`, the old implementation made that `[ | ] H` was introducing `H : n=0` in the first resulting goal but `H : nat` (the predecessor) in the second goal. On the contrary, the current (vanilla) design has the property that a list `p1 ... pn` of simple introduction patterns applies pointwise to the next `n` first quantified hypotheses or let-ins of the goal.

I don't know what is the best way to explain the question of completion. Maybe can it be said that if `[pat1 ... patn]` has less simple patterns than the number of arguments in the constructor, it automatically expands to `[pat1 ... patn ? ... ?]` with as many extra introduction patterns so as to match the arity of the constructor. Note that let-ins are included in the count.

Reviewer: Yes, I think the explanation you give in the last paragraph is a good one.
@jfehrle Could you deprecate the `Bracketing Last Introduction Pattern` as part of this PR?

   .. flag:: Bracketing Last Introduction Pattern

      Force completion, if needed, when the last introduction pattern is a
      disjunctive or conjunctive pattern (on by default).

.. note::

   :n:`intros {+ p}` is not equivalent to :n:`intros p; ... ; intros p`
   for the following reason: If one of the :n:`p` is a wildcard pattern, it
   might succeed in the first case because the further hypotheses it
   depends on are eventually erased too while it might fail in the second
   case because of dependencies in hypotheses which are not yet
   introduced (and a fortiori not yet erased).

Examples:

.. _intropattern_conj_ex:

>    .. example:: intro pattern for /\\

Author: The text on the right appears in the header for the example in the output.  Seems useful

  .. exn:: Not the right number of missing arguments.
     :undocumented:

.. _intropatterns:

Intro patterns
~~~~~~~~~~~~~~

Intro patterns let you specify the name to assign to variables and hypotheses
introduced by tactics.  They also let you split an introduced hypothesis into
multiple hypotheses or subgoals.  Common tactics that accept intro patterns
include :tacn:`apply`, :tacn:`intros` and :tacn:`destruct`.

.. productionlist::        coq
   ne_intropatterns        : `intropattern` ... `intropattern`
   intropatterns           : `intropattern` ... `intropattern`
                           : `empty`

>    empty                   :

Reviewer: I like a lot this way you found to make the empty case explicit, and it is also clear why this is needed. However, I would avoid repeating the ``` `intropattern` ... `intropattern` ``` part. So I suggest changing the above to:

```
    intropatterns           : `ne_intropatterns`
                            : `empty`
    ne_intropatterns        : `intropattern` ... `intropattern`
    empty                   :

   intropattern            : *
                           : **
                           : `simple_intropattern`
   simple_intropattern     : `simple_intropattern_closed` % `term` ... % `term`
   simple_intropattern_closed : `naming_intropattern`
                              : _
                              : `or_and_intropattern`
                              : `equality_intropattern`
   naming_intropattern     : `ident`
                           : ?
                           : ?`ident`
   or_and_intropattern     : [ `intropatterns` | ... | `intropatterns` ]

>                            : ()

Reviewer: I don't know this intro pattern and it doesn't look like it is explained below.

Reviewer: Ping: this intro-pattern should be removed from the grammar.

                           : ( `simple_intropattern` , ... , `simple_intropattern` )
                           : ( `simple_intropattern` & ... & `simple_intropattern` )
   equality_intropattern   : ->
                           : <-
                           : [= `intropatterns` ]

Use these elementary patterns to specify a name:

* :n:`@ident` - use the specified name
* :n:`?` - let Coq choose a name
* :n:`?@ident` - generate a name that begins with :n:`@ident`
* :n:`_` - discards the matched part (unless it is required for another
  hypothesis)

> * if a disjunction pattern omits a name, such as :g:`[|H2]`, Coq will choose a name

Reviewer: A conjunctive or disjunctive pattern. In fact the simplest (and most useful) example is probably the common `[]`.

Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,
use the :tacn:`split` tactic to replace the current goal with subgoals :g:`A` and :g:`B`.
For a goal :g:`A \/ B`, use :tacn:`left` to replace the current goal with :g:`A`, or
:tacn:`right` to replace the current goal with :g:`B`.

**describe [] and , forms, mention they are used but not recommended**

* :n:`( {+, @simple_intropattern}` ) - matches
  a product over an inductive type :g:`I` with a single
  constructor that itself has at least `n` arguments.  It performs a case
  analysis over the hypothesis, as :n:`@destruct` would, and applies the
  patterns to the arguments of the constructor of :g:`I`.  Note that
  that :n:`( {+, @simple_intropattern} )` is equivalent to :n:`[{+ @simple_intropattern}]`.

**need example**

* :n:`( {+& @simple_intropattern} )` - a shortcut for matching with
  :n:`(@simple_intropattern,( ... ,( ..., @simple_intropattern ) ... ))`.  It expects the hypothesis to be a sequence of
  right-associative binary inductive constructors such as :g:`conj` or
  :g:`ex_intro`.
  :ref:`Example <intropattern_ampersand_ex>`

> The other splitting patterns described below are not frequently used.

Reviewer: In the current form, I still disagree with this sentence. For instance, I do use *very* frequently the `->` / `<-` and `*` intro patterns.

  applies either :tacn:`injection` or :tacn:`discriminate`
  instead of :tacn:`destruct`; if :tacn:`injection` is applicable, the intropatterns
  are used on the hypotheses generated by :tacn:`injection`; if the
  number of patterns is smaller than the number of hypotheses generated, the
  pattern :n:`?` is used to complete the list.
  :ref:`Example <intropattern_injection_ex>`

* :n:`*` - introduces one or more quantified variables from the result
  until there are no more quantified variables.
  :ref:`Example <intropattern_star_ex>`

* :n:`**` - introduces one or more quantified variables or hypotheses from the result until there are
  no more quantified variables or implications (:g:`->`).  :g:`intros **` is equivalent
  to :g:`intros`.
  :ref:`Example <intropattern_2stars_ex>`

* :n:`@simple_intropattern_closed {* % @term}` - first applies each of the terms
  with the :tacn:`apply` tactic on the hypothesis to be introduced, then it applies :n:`@simple_intropattern_closed`.
  :ref:`Example <intropattern_injection_ex>`

  the on-the-fly application of lemmas: :n:`p{+ % @term}` where :n:`p`
  itself is not a pattern for on-the-fly application of lemmas (note:

>   syntax is in experimental stage)  **huh?**

Author: Will do.  FWIW: Better to say "Anyway" here.  I don't recall ever using "Anyways", even in casual speech.

```
“Anyways” with the S is considered slang, and is a part of nonstandard, colloquial, or
informal English. Furthermore, since “anyway” is an adverb and it is impossible for
adverbs to be plural. In newspapers or in any example of formal writing, you will
almost never see the word “anyways” used.
```

>   syntax is in experimental stage)  **huh?**

Reviewer: Interesting, I had read somewhere else that the two were equivalent...

  .. exn:: Not the right number of missing arguments.
     :undocumented:

.. _intropatterns:

Intro patterns
~~~~~~~~~~~~~~

Intro patterns let you specify the name to assign to variables and hypotheses
introduced by tactics.  They also let you split an introduced hypothesis into
multiple hypotheses or subgoals.  Common tactics that accept intro patterns

> include :tacn:`apply`, :tacn:`intros` and :tacn:`destruct`.

Reviewer: It's not `apply` but `apply ... in` which is a very different (and much less known) variant which supports an `as` clause. Thus, I would not cite `apply` in this list. I would rather cite `assert` or `specialize`.

Author: I see your point.  My idea for revising doc for `apply` and other tactics is to show the full syntax covering all options first, then identifying the most common options, followed by the details, akin to this section.  So there may only be a single index entry for `apply`.  I'll put in `assert`.

.. productionlist::        coq
   intropatterns           : `intropattern` ... `intropattern`
                           : `empty`
   empty                   :
   intropattern            : *
                           : **
                           : `simple_intropattern`
   simple_intropattern     : `simple_intropattern_closed` [ % `term` ... % `term` ]
   simple_intropattern_closed : `naming_intropattern`
                              : _
                              : `or_and_intropattern`
                              : `equality_intropattern`
   naming_intropattern     : `ident`
                           : ?
                           : ?`ident`
   or_and_intropattern     : [ `intropatterns` | ... | `intropatterns` ]
                           : ()
                           : ( `simple_intropattern` , ... , `simple_intropattern` )
                           : ( `simple_intropattern` & ... & `simple_intropattern` )
   equality_intropattern   : ->
                           : <-
                           : [= `intropatterns` ]

>    or_and_intropattern_loc : or_and_intropattern

Reviewer: ```suggestion
   or_and_intropattern_loc : `or_and_intropattern`
```

                           : `ident`

Note that the intro pattern syntax varies between tactics.
Most tactics use :n:`@simple_intropattern` in the grammar.
:tacn:`destruct`, :tacn:`edestruct`, :tacn:`induction`,
:tacn:`einduction`, :tacn:`case`, :tacn:`ecase` and the various
:tacn:`inversion` tactics use :n:`@or_and_intropattern_loc`, while
:tacn:`intros` and :tacn:`eintros` use :n:`@intropatterns`.
The :n:`eqn:` construct in various tactics uses :n:`@naming_intropattern`.

**Naming patterns**

Use these elementary patterns to specify a name:

* :n:`@ident` - use the specified name
* :n:`?` - let Coq choose a name
* :n:`?@ident` - generate a name that begins with :n:`@ident`
* :n:`_` - discards the matched part (unless it is required for another
  hypothesis)
* if a disjunction pattern omits a name, such as :g:`[|H2]`, Coq will choose a name

**Splitting patterns**

The most common splitting patterns are:

* split a hypothesis in the form :n:`A /\ B` into two
  hypotheses :g:`H1: A` and :g:`H2: B` using the pattern :g:`(H1 & H2)` or

>   :g:`(H1, H2)` or :g:`[H1 H2]`

Reviewer: ```suggestion
  :g:`(H1, H2)` or :g:`[H1 H2]`.
```

Author: FWIW: Technically, this is a sentence fragment (lacking a subject), therefore it doesn't really need a period at the end, although in the second bullet point I put a period to separate the fragment from the sentence that follows up.

  :ref:`Example <intropattern_conj_ex>`
* split a hypothesis in the form :g:`A \/ B` into two
  subgoals using the pattern :g:`[H1|H2]`.  The first subgoal will have the hypothesis
  :g:`H1: A` and the second subgoal will have the hypothesis :g:`H2: B`.
  :ref:`Example <intropattern_disj_ex>`
* split a hypothesis in either of the forms :g:`A /\ B` or :g:`A \/ B` using the pattern :g:`[]`.

> **A node has to be of type Prop to be matched/split/named, right?**

Reviewer: No.

**what other common operators are intro patterns useful for?  ->? <->? forall?  I would like to

> mention them**

Reviewer: Yes `<->`: use a conjonctive intro-pattern as with `/\`. `forall` and `->` obviously but because intro patterns are used to introduce variables / hypotheses in the context, that's all.

Author: Ah, I should have said "which splitting patterns are useful for other operators?".  <-> splits nicely but I suppose `forall` does not.

Patterns can be nested: :n:`[[Ha|Hb] H]` can be used to split :n:`(A \/ B) /\ C`.

> Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,

Author: Strictly speaking, a goal is the hypotheses plus the conclusion, right?  The same as "proof context"?  "proof context" or "context" seems clearer than "goal".  References such as "For a goal :g:`A /\ B`" on this line seem very natural.  Or do we want to be strict and always use the term"conclusion"?

> Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,

Reviewer: Hum. Strictly speaking, we may include the context in what we call "a goal" (and this is what `Show Goal` does). But when we speak more specifically of what appears below the bar, we generally say "goal" as well. Here, I meant the conclusion of the latter, i.e. what goes after the last `->`

> Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,

Author: Huh?  Where is the `->` here?  "conclusion" seems out of place.  I'm talking about applying `split` in this case:

```
1 subgoal
A, B : Prop
______________________________________(1/1)
A /\ B
```
The paragraph is not about intro patterns per se; it's meant to provide clarifying contrast for when you have `A /\ B` below the line.  Perhaps it would be better if the first sentence was "Note that you must use other tactics to split up goals."

Or I could drop the paragraph entirely.

> Note that there is no equivalent to intro patterns for goals.  For a goal :g:`A /\ B`,

Reviewer: The conclusion is the full goal because there is no `->`. Anyway, it seems that the "clarification" I proposed wasn't one in reality. I prefer that we keep your paragraph as is. I think it is really useful.

use the :tacn:`split` tactic to replace the current goal with subgoals :g:`A` and :g:`B`.
For a goal :g:`A \/ B`, use :tacn:`left` to replace the current goal with :g:`A`, or
:tacn:`right` to replace the current goal with :g:`B`.

* **() - I couldn't figure out how this worked.  I think it did something for intros (lost the example) but
  it wasn't valid for any destruct I tried.  Is it actually useful?**

* :n:`( {+, @simple_intropattern}` ) - matches
  a product over an inductive type with a
  :ref:`single constructor <intropattern_cons_note>`.
  If the number of patterns
  equals the number of constructor arguments, then it applies the patterns only to
  the arguments, and
  :n:`( {+, @simple_intropattern} )` is equivalent to :n:`[{+ @simple_intropattern}]`.
  If the number of patterns equals the number of constructor arguments plus the number
  of :n:`let-ins`, the patterns are applied to the arguments and :n:`let-in` variables.

* :n:`( {+& @simple_intropattern} )` - matches a term that consists of one or more

>   right-hand nested binary inductive types in a form such as :g:`(a1 OP (a2 OP ...))`.

Author: OK.  Though generally "right-associative" is used to refer to a property of the operator rather than how the expression is nested.

>   right-hand nested binary inductive types in a form such as :g:`(a1 OP (a2 OP ...))`.

Reviewer: Oh, right. It's just that the usual case is with non-parenthesized expressions, in which case it will be nested that way only if the operators are right-associative. But your previous sentence made sense, indeed!

  The splitting pattern can have more than 2 names, for example :g:`(H1 & H2 & H3)`
  matches :g:`A /\ B /\ C`.
  The inductive types must have a
  :ref:`single constructor with two parameters <intropattern_cons_note>`.
  :ref:`Example <intropattern_ampersand_ex>`

* :n:`[ {+| @intropatterns} ]` - splits an inductive type that has
  :ref:`multiple constructors <intropattern_cons_note>`
  such as :n:`A \/ B`

>   into multiple subgoals.  The number of ``intropatterns`` must be the same as the number of

Reviewer: ```suggestion
  into multiple subgoals.  The number of :token:`intropatterns` must be the same as the number of
```

Author: FWIW It's awkward to have both `intropattern` and `intropatterns` non-terminals--makes it hard to distinguish singular and plural references.

Author: I think we should banish plural non-terminal names like `intropatterns` in favor of `intropattern_list`.  But not in this PR.

Reviewer: Well, I agree at least that as part of this PR, the token be named `introppatern_list` instead of `intropatterns`.

Author: Done, but only in the doc.

  constructors for the matched part.

> * :n:`[ {* @intropattern} ]` - splits an inductive type that has a

Reviewer: I think you meant ``` :n:`[ @intropattern ]` ``` here. But I would say that you can simply remove this bullet point and change the `+` to `*` in the previous bullet point as the explanation above applies to this case as well (understanding "multiple" in the large sense) and anyways the `( , )` intro-pattern is recommended for this case.

Author: Ah, no.  `[]` works for both A /\ B and A \\/ B, whereas the 2 prior bullets are specific to one or the other.  I will add "(a special case of the last 2 bullets)".

(Where are the recommendations from?  And why does it matter to the user?  Both forms are valid and AFAIK we're not planning to remove one of them.]

Reviewer: The disjunctive form is the most general because the conjunctive form is just its specialization to inductive types with a single constructor. This is what I meant.
OTOH, `[]` is very special.

  :ref:`single constructor with multiple parameters <intropattern_cons_note>`
  such as :n:`A /\ B` into multiple hypotheses.  Use :n:`[H1 [H2 H3]]` to match :g:`A /\ B /\ C`.
* :n:`[]` - splits an inductive type:  If the inductive type has multiple constructors,
  create one subgoal for each constructor.  If the inductive type has a single constructor with
  multiple parameters, split it into multiple hypotheses.

**it's split into multiple hyps when the inductive type is (A B C:Prop) A -> B -> C -> something
but probably not for terms with other types, e.g. nat.  Is this behavior tied to the intro pattern or more

> tied to ->?**

Reviewer: I did not understand this question. Can you reformulate?

**Equality patterns**

These patterns can be used when the hypothesis is an equality:

* :n:`->` - replaces the right-hand side of the hypothesis with the left-hand
  side of the hypothesis in the conclusion of the goal; the hypothesis is
  cleared; if the left-hand side of the hypothesis is a variable, it is
  substituted everywhere in the context and the variable is removed.
  :ref:`Example <intropattern_rarrow_ex>`
* :n:`<-` - similar to :n:`->`, but replaces the left-hand side of the hypothesis
  with the right-hand side of the hypothesis.
* :n:`[= {*, @intropattern} ]` - If the product is over an equality type,
  applies either :tacn:`injection` or :tacn:`discriminate`

>   instead of :tacn:`destruct`; if :tacn:`injection` is applicable, the intropatterns

Author: Good point, leftover from the old text.

  number of patterns is smaller than the number of hypotheses generated, the
  pattern :n:`?` is used to complete the list.
  :ref:`Example <intropattern_inj_discr_ex>`

**Other patterns**

* :n:`*` - introduces one or more quantified variables from the result
  until there are no more quantified variables.
  :ref:`Example <intropattern_star_ex>`

* :n:`**` - introduces one or more quantified variables or hypotheses from the result until there are
  no more quantified variables or implications (:g:`->`).  :g:`intros **` is equivalent
  to :g:`intros`.
  :ref:`Example <intropattern_2stars_ex>`

* :n:`@simple_intropattern_closed {* % @term}` - first applies each of the terms
  with the :tacn:`apply` tactic on the hypothesis to be introduced, then it uses
  :n:`@simple_intropattern_closed`.
  :ref:`Example <intropattern_injection_ex>`

.. flag:: Bracketing Last Introduction Pattern

   For :n:`intros @intro_pattern_list`, controls how to handle a
   conjunctive pattern that doesn't give enough simple patterns to match
   all the arguments in the constructor.  If set (the default), |Coq| generates
   additional names to match the number of arguments.

>    Unsetting the option will _____________, behavior that is more

Author: "goal", as you put it, is better than "conclusion" here, eh?

>    Unsetting the option will _____________, behavior that is more

Reviewer: Indeed, but when I was recommending the use of "conclusion", this is not what I meant (as explained above): the conclusion is the part of the goal that is after the last `->`.

   similar to |SSR|'s ``intro`` patterns.

.. _intropattern_cons_note:

.. note::

   :n:`A \/ B` and :n:`A /\ B` use infix notation to refer to the inductive
   types :n:`or` and :n:`and`.
   :n:`or` has multiple constructors (:n:`or_introl` and :n:`or_intror`),
   while :n:`and` has a single constructor (:n:`conj`) with multiple parameters
   (:n:`A` and :n:`B`).
   These are defined in theories/Init/Logic.v.  The "where" clauses define the
   infix notation for "or" and "and".

   | Inductive or (A B:Prop) : Prop :=
   |   | or_introl : A -> A \\/ B
   |   | or_intror : B -> A \\/ B
   | where "A \\/ B" := (or A B) : type_scope.
   |
   | Inductive and (A B:Prop) : Prop :=
   |   conj : A -> B -> A /\\ B
   | where "A /\\ B" := (and A B) : type_scope.

>    **quoted block above isn't formatted as I wanted.  What's the right way to get output to exactly match the input?**

Author: I had tried `::` and indentation but Sphinx seemed to get confused by the indented lines within.  `coqdoc` works better, although it gives an unneed blank line at the top.

         destruct H as (HA & HB).

.. _intropattern_disj_ex:

   .. example:: intro pattern for \\/

      .. coqtop:: reset none

         Goal forall (A: Prop) (B: Prop), (A \/ B) -> True.

      .. coqtop:: out

         intros.

      .. coqtop:: all

         destruct H as [HA|HB]. all: swap 1 2.

.. _intropattern_rarrow_ex:

   .. example:: -> intro pattern

      .. coqtop:: reset none

         Goal forall (x:nat) (y:nat) (z:nat), (x = y) -> (y = z) -> (x = z).

      .. coqtop:: out

         intros * H.

      .. coqtop:: all

         intros ->.

.. _intropattern_inj_discr_ex:

   .. example:: [=] intro pattern

      The first :n:`intros [=]` uses :tacn:`injection` to strip :n:`(S ...)` from
      both sides of the matched equality.  The second uses :tacn:`discriminate` on
      the contradiction :n:`1 = 2` (internally represented as :n:`(S O) = (S (S O))`)
      to complete the goal.

      .. coqtop:: reset none

         Goal forall (n m:nat),  (S n) = (S m) -> (S O)=(S (S O)) -> False.

      .. coqtop:: out

         intros *.

      .. coqtop:: all

         intros [= H].

      .. coqtop:: all

         intros [=].

.. _intropattern_injection_ex:

>    .. example:: [=->%length_zero_iff_nil] intro pattern

Reviewer: I would call this example “Advanced combination illustrating `->`, `%` and `[=]`” and probably put it at the end.

>       .. from https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v

Reviewer: I suggest you put the credits in the text of the example.

      | :g:`Theorem length_zero_iff_nil (l : list A):`

>       |    :g:`length l = 0 <-> l=[].`

Reviewer: Use `.. coqdoc::`, or `Check` inside a `.. coqtop::` block

>       * `intros` would add :g:`H : S (length ys) = 1`

Reviewer: The ordering is weird. It would be natural if this were commenting on the code above but this is commenting on the text below. What about reversing completely the order in this example: first the code, then the bullet points, then the theorem statement.

      * `intros [=]` would additionally apply :tacn:`injection` to :g:`H` to yield :g:`H0 : length ys = 0`
      * `intros [=->%length_zero_iff_nil]` applies the theorem, making H the equality :g:`l=[]`,

>         which is then applied as for :g:`->`.  (Note :g:`nil` is the same as :g:`[]`.)

Reviewer: Rather than adding this note, you could make the theorem statement talk about `nil` instead of using the notation.

      .. coqtop:: reset none

         Require Import Coq.Lists.List.
         Section IntroPatterns.
         Variables (A : Type) (xs ys : list A).

      .. coqtop:: out

         Example ThreeIntroPatternsCombined :
         S (length ys) = 1 -> xs ++ ys = xs.

      .. coqtop:: all

         intros [=->%length_zero_iff_nil].

.. _intropattern_ampersand_ex:

   .. example:: (A & B & ...) intro pattern

      .. coqtop:: reset none

>          Goal forall (A: Prop) (B: nat -> Prop) (C: Prop), A /\ (exists x:nat, B x /\ C) -> True.

Reviewer: I suggest putting the parameters in a `Variables` listing as in the previous example so that they don't pollute the context.

* split a hypothesis in the form :n:`A /\ B` into two
  hypotheses :g:`H1: A` and :g:`H2: B` using the pattern :g:`(H1 & H2)` or
  :g:`(H1, H2)` or :g:`[H1 H2]`.
  :ref:`Example <intropattern_conj_ex>`.  This also works on :n:`A <-> B`, which

>   is just a notation representing :n:`(A -> B) /\ (B -> A)`.

Reviewer: Correct. But note that there were discussions of creating a proper inductive type for `<->` in which case this splitting pattern would still work. That's the magic :smile:

  of :n:`let-ins`, the patterns are applied to the arguments and :n:`let-in` variables.

* :n:`( {+& @simple_intropattern} )` - matches a term that consists of one or more

>   nested binary inductive types in a right-associative form such as :g:`(a1 OP (a2 OP ...))`.

Reviewer: Maybe keep your old formulation but instead remove the parentheses from your example and add "(supposing that OP is right-associative)"...

  The splitting pattern can have more than 2 names, for example :g:`(H1 & H2 & H3)`
  matches :g:`A /\ B /\ C`.
  The inductive types must have a
  :ref:`single constructor with two parameters <intropattern_cons_note>`.
  :ref:`Example <intropattern_ampersand_ex>`

> * :n:`[ {*| @intropatterns} ]` - splits an inductive type that has

Reviewer: Here we go. Now that you have changed `+` in `*`, the bullet-point above is just a specialization of this case.

Author: Not quite.  Both the 2 preceding points should have `+`.

  :ref:`multiple constructors <intropattern_cons_note>`
  such as :n:`A \/ B`
  into multiple subgoals.  The number of :token:`intropatterns` must be the same as the number of
  constructors for the matched part.
* :n:`[ {* @intropattern} ]` - splits an inductive type that has a
  :ref:`single constructor with multiple parameters <intropattern_cons_note>`
  such as :n:`A /\ B` into multiple hypotheses.  Use :n:`[H1 [H2 H3]]` to match :g:`A /\ B /\ C`.

> * :n:`[]` - splits an inductive type (a special case of the last 2 bullets):  If the inductive

Reviewer: On the contrary, I would not say that it is a "special case". It has a peculiar semantics...

         intros * [a | (_,c)] f.
         all: swap 1 2.

.. _intropattern_injection_ex:

   .. example:: Combination intro pattern using [=], -> and %

      .. coqtop:: reset none

         Require Import Coq.Lists.List.
         Section IntroPatterns.
         Variables (A : Type) (xs ys : list A).

      .. coqtop:: out

         Example ThreeIntroPatternsCombined :
         S (length ys) = 1 -> xs ++ ys = xs.

      .. coqtop:: all

         intros [=->%length_zero_iff_nil].

      * `intros` would add :g:`H : S (length ys) = 1`
      * `intros [=]` would additionally apply :tacn:`injection` to :g:`H` to yield :g:`H0 : length ys = 0`
      * `intros [=->%length_zero_iff_nil]` applies the theorem, making H the equality :g:`l=nil`,
        which is then applied as for :g:`->`.

      .. coqdoc::

         Theorem length_zero_iff_nil (l : list A):
            length l = 0 <-> l=nil.

>       The example is based on https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v

Reviewer: Very last comment: I suggest using a permalink, like https://github.com/tchajed/coq-tricks/blob/8e6efe4971ed828ac8bdb5512c1f615d7d62691e/src/IntroPatterns.v. Since it's a bit long, maybe this can become ``` `Tej Chajed's coq-tricks
<https://github.com/tchajed/coq-tricks/blob/8e6efe4971ed828ac8bdb5512c1f615d7d62691e/src/IntroPatterns.v>`_ ```
